audio/src/soundio.zig

```diff
@@ -0,0 +1,201 @@
+const std = @import("std");
+const mem = std.mem;
+const fmt = std.fmt;
+const c = @import("soundio").c;
+const SoundIo = @import("soundio").SoundIo;
+const SoundIoDevice = @import("soundio").Device;
+const SoundIoInStream = @import("soundio").InStream;
+const SoundIoOutStream = @import("soundio").OutStream;
+const Mode = @import("main.zig").Mode;
+const DeviceDescriptor = @import("main.zig").DeviceDescriptor;
+
+const is_raw_sign = "MD_IS_RAW";
+const input_mode_sign = "MD_INPUT_";
+const output_mode_sign = "MD_OUTPUT";
+
+const Audio = @This();
+
+const DeviceUnion = union(Mode) {
+    input: SoundIoInStream,
+    output: SoundIoOutStream,
+};
+
+pub const DeviceDescriptorInternal = SoundIoDevice;
+pub const Device = struct {
+    handle: DeviceUnion,
+
+    pub fn deinit(self: Device) void {
+        return switch (self.handle) {
+            .input => |d| d.deinit(),
+            .output => |d| d.deinit(),
+        };
+    }
+};
+pub const DeviceIterator = struct {
+    ctx: Audio,
+    mode: Mode,
+    device_len: u16,
+    index: u16,
+
+    pub fn next(self: *DeviceIterator) IteratorError!?DeviceDescriptor {
+        if (self.index < self.device_len) {
+            const device_desc = switch (self.mode) {
+                .input => self.ctx.handle.getInputDevice(self.index) orelse return null,
+                .output => self.ctx.handle.getOutputDevice(self.index) orelse return null,
+            };
+            self.index += 1;
+            return DeviceDescriptor{
+                .id = try fmt.allocPrint(
+                    self.ctx.allocator,
+                    "{s}[{s}][{s}]",
+                    .{
+                        device_desc.id(),
+                        is_raw_sign,
+                        switch (@intToEnum(Mode, device_desc.handle.aim)) {
+                            .input => input_mode_sign,
+                            .output => output_mode_sign,
+                        },
+                    },
+                ),
+                .name = device_desc.name(),
+                .internal = device_desc,
```

Comment:
Lets not store it, we should get the device only in ``requestDevice`` as mentioned above.